File: 실버 1 (3문제)/11660_구간 합 구하기 5.py
# 설명 링크 : https://moon-gd.tistory.com/73
import sys
q_input = sys.stdin.readline # 빠른 입력을 위해서

# 정사각형 크기와 답 출력 횟수 입력
size, times = map(int, q_input().split())

# 누적 합 배열 저장
sums = [[0 for _ in range(size + 1)] for _ in range(size + 1)]

for i in range(1, size + 1):
    numbers = list(map(int, q_input().split()))

    # 누적 합 계산
    for j in range(1, size + 1):
        sums[i][j] = sums[i - 1][j] + sums[i][j-1] - sums[i - 1][j - 1] + numbers[j - 1]


for _ in range(times):
    start_row, start_col, end_row, end_col = map(int, q_input().split())

    print(
        sums[end_row][end_col]
        - sums[start_row - 1][end_col] - sums[end_row][start_col - 1]
        + sums[start_row - 1][start_col - 1]
    )

File: 실버 1 (7문제)/11660_구간 합 구하기 5.c
// 설명 링크 : https://moon-gd.tistory.com/73
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#define MAX 1025

int number;
int sums[MAX][MAX];  // 누적 합

int main() {
	int size = 0, times = 0;
	scanf("%d %d", &size, &times);

	int i, j;
	for (i = 1; i <= size; i++) {
		for (j = 1; j <= size; j++) {
			scanf("%d", &number);

			// 누적 합 정보 저장
			sums[i][j] = sums[i][j - 1] + sums[i - 1][j] - sums[i - 1][j - 1] + number;
		}
	}

	int start_row, start_col, end_row, end_col;

	for (i = 1; i <= times; i++) {
		scanf("%d %d %d %d", &start_row, &start_col, &end_row, &end_col);

		printf("%d\n", sums[end_row][end_col] - sums[start_row - 1][end_col] - sums[end_row][start_col - 1] + sums[start_row - 1][start_col - 1]);
	}

	return 0;
}